### PR TITLE
Add better type definition

### LINKED ISSUE #60: How to use with Typescript?
What's the recommended way to use this currently with Typescript as all the examples are broken for me in the latest vscode.

Typescript: `3.9.0-dev-20200408`
VS Code: `1.44.0`
Camelcase-keys: `6.2.2`

```ts
import camelcaseKeys from 'camelcase-keys';

// Convert an object
camelcaseKeys({ 'foo-bar': true });
// (alias) camelcaseKeys<{
//     'foo-bar': boolean;
// }>(input: {
//     'foo-bar': boolean;
// }, options ?: camelcaseKeys.Options): {
//     'foo-bar': boolean;
// } (+1 overload)

// Convert an array of objects
camelcaseKeys([{ 'foo-bar': true }, { 'bar-foo': false }]);
// (alias) camelcaseKeys<({
//     'foo-bar': boolean;
//     'bar-foo'?: undefined;
// } | {
//     'bar-foo': boolean;
//     'foo-bar'?: undefined;
// })[]>(input: ({
//     'foo-bar': boolean;
//     'bar-foo'?: undefined;
// } | {
//     'bar-foo': boolean;
//     'foo-bar'?: undefined;
// })[], options ?: camelcaseKeys.Options): ({
//     ...;
// } | {
//     ...;
// })[](+1 overload)

camelcaseKeys({ 'foo-bar': true, nested: { unicorn_rainbow: true } }, { deep: true });
// (alias) camelcaseKeys<{
//     'foo-bar': boolean;
//     nested: {
//         unicorn_rainbow: boolean;
//     };
// }>(input: {
//     'foo-bar': boolean;
//     nested: {
//         unicorn_rainbow: boolean;
//     };
// }, options ?: camelcaseKeys.Options): {
//     'foo-bar': boolean;
//     nested: {
//         ...;
//     };
// } (+1 overload)

camelcaseKeys({ a_b: 1, a_c: { c_d: 1, c_e: { e_f: 1 } } }, { deep: true, stopPaths: ['a_c.c_e'] }),
// (alias) camelcaseKeys<{
//     a_b: number;
//     a_c: {
//         c_d: number;
//         c_e: {
//             e_f: number;
//         };
//     };
// }>(input: {
//     a_b: number;
//     a_c: {
//         c_d: number;
//         c_e: {
//             e_f: number;
//         };
//     };
// }, options?: camelcaseKeys.Options): {
//     a_b: number;
//     a_c: {
//         c_d: number;
//         c_e: {
//             ...;
//         };
//     };
// } (+1 overload)

// Convert object keys to pascal case
camelcaseKeys({ 'foo-bar': true, nested: { unicorn_rainbow: true } }, { deep: true, pascalCase: true });
// (alias) camelcaseKeys<{
//     'foo-bar': boolean;
//     nested: {
//         unicorn_rainbow: boolean;
//     };
// }>(input: {
//     'foo-bar': boolean;
//     nested: {
//         unicorn_rainbow: boolean;
//     };
// }, options ?: camelcaseKeys.Options): {
//     'foo-bar': boolean;
//     nested: {
//         ...;
//     };
// } (+1 overload)
```

### PR DESCRIPTION
**Pull request by g-plane**
This PR leverages TypeScript 4.1 template literal types to enhance type definition.

Fixes #60

### CODE REVIEW

**Comment by sindresorhus** [comment]
See https://github.com/sindresorhus/camelcase-keys/pull/69 and the feedback given there.

**Comment by g-plane** [comment]
Should we drop Node.js 8 in CI?

**Comment by sindresorhus** [comment]
Sure. I plan to target Node.js 12 after this PR.

**Comment by g-plane** [comment]
And, can you review this PR again?

File: index.d.ts [hunk 1]
```diff
@@ -1,3 +1,55 @@
+import {CamelCase, PascalCase} from 'type-fest';
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+type EmptyTuple = [];
+
+type IsInclude<List extends readonly unknown[] | undefined, Target> =
```
Comment (sindresorhus) [inline index.d.ts:6]: These types need a short doc comment about what they do.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
> Should we drop Node.js 8 in CI?

Yes

File: index.test-d.ts [hunk 2]
```diff
@@ -3,33 +3,57 @@ import camelcaseKeys = require('.');
 
 const fooBarObject = {'foo-bar': true};
 const camelFooBarObject = camelcaseKeys(fooBarObject);
-expectType<typeof fooBarObject>(camelFooBarObject);
+expectType<{ fooBar: boolean }>(camelFooBarObject);
```
Comment (sindresorhus) [inline index.test-d.ts:6]:
```suggestion
expectType<{fooBar: boolean}>(camelFooBarObject);
```

File: index.test-d.ts [hunk 3]
```diff
@@ -3,33 +3,57 @@ import camelcaseKeys = require('.');
...
+	camelcaseKeys(
+		{'foo-bar': true, foo_bar: true},
+		{exclude: ['foo', 'foo_bar', /bar/] as const}
+	)
 );
 
-expectType<{[key in 'foo-bar']: true}>(
+expectType<{[key in 'fooBar']: boolean}>(
 	camelcaseKeys({'foo-bar': true}, {stopPaths: ['foo']})
 );
```
Comment (sindresorhus) [inline index.test-d.ts:58]: Can you look at the tests in https://github.com/sindresorhus/camelcase-keys/pull/69/files#diff-724a6378efca7770e513d0ca30555fe214930f896214ac52213320a0b354df48R1 and see if there's anything there that could be added here too?

File: index.test-d.ts [hunk 4]
```diff
@@ -3,33 +3,57 @@ import camelcaseKeys = require('.');
...
+		{'foo-bar': {foo_bar: {'foo bar': true}}},
+		{deep: true}
+	)
 );
 
-expectType<{[key in 'foo-bar']: true}>(
-	camelcaseKeys({'foo-bar': true}, {deep: true, pascalCase: true})
+expectType<{[key in 'FooBar']: boolean}>(
+	camelcaseKeys({'foo-bar': true}, {pascalCase: true})
+);
```
Comment (sindresorhus) [inline index.test-d.ts:32]:
```suggestion
expectType<{FooBar: boolean}>(
	camelcaseKeys({'foo-bar': true} as const, {pascalCase: true})
);
```

?

I saw it from the other PRs tests.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
How does this implementation differ from https://github.com/sindresorhus/camelcase-keys/pull/69 ? I'm trying to make sure this PR is as good or better has that one. I haven't reviewed this in detail yet.

**Comment by g-plane** [comment]
Both are similar, but:
- This PR moves some utilities types out of `camelcaseKeys` namespace.
- This PR doesn't handle `stopPaths`, because TypeScript type definition isn't imperative.
- #69 is better than this PR when handling tuples. I may learn #69 then improve this PR.

File: readme.md [hunk 5]
```diff
@@ -63,13 +63,17 @@ Default: `[]`
 
 Exclude keys from being camel-cased.
 
+If you're using TypeScript and this option can be statically determined, it's recommended that adding `as const` for it.
```
Comment (sindresorhus) [inline readme.md:66]: This is better documented in index.d.ts

File: index.d.ts [hunk 6]
```diff
@@ -1,3 +1,83 @@
...
+ */
+type CamelCaseKeys<
+	T extends Record<string, any> | readonly any[],
+	Deep extends boolean,
+	IsPascalCase extends boolean,
+	Exclude extends readonly unknown[],
+	StopPaths extends readonly string[],
+	Path extends string = ''
+> = T extends readonly any[]
+	? {
```
Comment (sindresorhus) [inline index.d.ts:47]: The whole method here could use some short inline code comments. It's quite hard to follow.

File: index.d.ts [hunk 7]
```diff
@@ -1,3 +1,83 @@
...
+
+/**
+ * Append a segment to dot-notation path.
+ */
+type AppendPath<S extends string, Last extends string> = S extends ''
+	? Last
+	: `${S}.${Last}`;
+
+/**
+ * Convert keys of an object to camelcase strings.
```
Comment (sindresorhus) [inline index.d.ts:37]: Don't prefix with `*`
```suggestion
Convert keys of an object to camelcase strings.
```

File: index.d.ts [hunk 8]
```diff
@@ -1,3 +1,83 @@
+import {CamelCase, PascalCase} from 'type-fest';
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+type EmptyTuple = [];
+
+/**
+ * Return a default type if input type is nil.
+ *
+ * @template T input type
+ * @template U default type
```
Comment (sindresorhus) [inline index.d.ts:10]:
```suggestion
 * @template U - Default type.
```

File: index.test-d.ts [hunk 9]
```diff
@@ -3,35 +3,85 @@ import camelcaseKeys = require('.');
...
 
-expectType<Array<{[key in 'foo-bar']: true}>>(camelcaseKeys([{'foo-bar': true}]));
+expectType<Array<{fooBar: boolean}>>(camelcaseKeys([{'foo-bar': true}]));
 
 expectType<string[]>(camelcaseKeys(['name 1', 'name 2']));
 
 expectType<string[]>(camelcaseKeys(['name 1', 'name 2'], {deep: true}));
 
-expectType<{[key in 'foo-bar']: true}>(camelcaseKeys({'foo-bar': true}));
+expectType<readonly [{ fooBar: true }, {fooBaz: true}]>(
```
Comment (sindresorhus) [inline index.test-d.ts:18]:
```suggestion
expectType<readonly [{fooBar: true}, {fooBaz: true}]>(
```

File: index.test-d.ts [hunk 10]
```diff
@@ -3,35 +3,85 @@ import camelcaseKeys = require('.');
...
+	)
+);
+
+expectType<Record<string, string>>(
+	camelcaseKeys({} as Record<string, string>)
+);
+
+expectType<Record<string, string>>(
+	camelcaseKeys({} as Record<string, string>, {deep: true})
+);
```
Comment (sindresorhus) [inline index.test-d.ts:85]: Are you sure the tests here cover all the cases?

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.test-d.ts (hunk 10 above)
Comment (g-plane) [reply index.test-d.ts:85]: Yes, I think, but you can suggest more cases.

**Comment by yuichkun** [comment]
Hi, I just wanted to stop by and say big thank you for implementing this! 👏 
Looking forward to leveraging this in my projects!
